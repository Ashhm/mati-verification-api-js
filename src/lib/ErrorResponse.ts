export default class ErrorResponse extends Error {
  constructor(message: string, public response: Response) {
    super(message);
  }
}
