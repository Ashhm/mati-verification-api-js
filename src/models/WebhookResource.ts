export enum EventNameTypes {
  StepCompleted = 'step_completed',
  VerificationCompleted = 'verification_completed',
  VerificationExpired = 'verification_expired',
  VerificationInputsCompleted = 'verification_inputs_completed',
  VerificationStarted = 'verification_started',
  VerificationUpdated = 'verification_updated',
}

export default interface WebhookResource {
  eventName: EventNameTypes;
  resource: string;
}
