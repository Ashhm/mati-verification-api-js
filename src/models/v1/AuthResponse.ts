export default interface AuthResponse {
  access_token: string;
  expires_in: number;
  token_type: 'Bearer';
}
