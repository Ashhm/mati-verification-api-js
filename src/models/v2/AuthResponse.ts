export interface UserPayload {
  _id: string;
}

export interface AuthPayload {
  user: UserPayload;
}

export default interface AuthResponse {
  access_token: string;
  expiresIn: number;
  payload: AuthPayload;
}
