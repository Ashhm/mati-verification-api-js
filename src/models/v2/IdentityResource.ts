export enum IdentityStatusTypes {
  Deleted = 'deleted',
  Pending = 'pending',
  Rejected = 'rejected',
  ReviewNeeded = 'reviewNeeded',
  Running = 'running',
  Verified = 'verified',
}

export default interface IdentityResource {
  _id: string; // TODO Rename
  flowId?: string;
  status: IdentityStatusTypes;
}
