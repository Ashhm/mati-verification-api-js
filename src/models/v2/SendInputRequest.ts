
export enum InputTypeTypes {
  DocumentPhoto = 'document-photo',
  SelfiePhoto = 'selfie-photo',
  SelfieVideo = 'selfie-video',
}

export enum DocumentTypeTypes {
  DrivingLicense = 'driving-license',
  NationalId = 'national-id',
  Passport = 'passport',
  ProofOfResidency = 'proof-of-residency',
}

export enum PageTypes {
  Back = 'back',
  Front = 'front',
}

export interface DocumentPhotoInputData {
  type: DocumentTypeTypes;
  country: string;
  region?: string;
  page: PageTypes;
  filename: string;
}

export interface Input<T> {
  inputType: InputTypeTypes,
  group: number,
  data: T, // TODO: TBD
}

export enum MediaTypeTypes {
  Document = 'document',
  Selfie = 'selfie',
  Video = 'video',
}

interface FileRecord {
  fileName: string;
  mediaType: MediaTypeTypes;
  stream: any;
}

export default interface SendInputRequest {
  inputs: Input<any>[],
  files: FileRecord[],
}
