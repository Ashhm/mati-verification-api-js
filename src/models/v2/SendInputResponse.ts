interface InputError {
  type: string;
  code: string;
  message?: string;
  details?: any;
}

interface InputSuccessResult {
  result: boolean;
}

interface InputErrorResult {
  error: InputError;
}

type InputResult = InputSuccessResult | InputErrorResult;

export default interface SendInputResponse extends Array<InputResult> {}
