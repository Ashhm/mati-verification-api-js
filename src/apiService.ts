import qs from 'qs';
import crypto from 'crypto';

import callHttp, { formContentType, RequestOptions } from './lib/callHttp';
import isomorph from './lib/isomorph';
import AuthResponse from './models/v2/AuthResponse';
import VerificationResource from './models/VerificationResource';
import IdentityResource from './models/v2/IdentityResource';
import SendInputRequest from './models/v2/SendInputRequest';
import SendInputResponse from './models/v2/SendInputResponse';
import ErrorResponse from './lib/ErrorResponse';

export const API_HOST = 'https://api.getmati.com/v2';

export type Options = {
  clientId: string;
  clientSecret: string;
  host?: string;
  webhookSecret?: string;
};

type CallHttpParamsType = {
  url?: string,
  path?: string,
  requestOptions?: RequestOptions,
  authType?: 'bearer' | 'basic' | 'none',
};

class ApiService {
  private bearerAuthHeader: string;

  private clientAuthHeader: string;

  private host: string;

  private webhookSecret: string | false = false;

  /**
   * Initializes the service.
   * @param options
   */
  init(options: Options) {
    const {
      host,
      clientId,
      clientSecret,
      webhookSecret,
    } = options;
    this.host = host || API_HOST;
    this.webhookSecret = webhookSecret || false;
    this.setClientAuth(clientId, clientSecret);
  }

  /**
   * Authenticates client.
   */
  async auth() {
    const authResponse = await this.callHttp({
      path: 'oauth',
      authType: 'basic',
      requestOptions: {
        method: 'POST',
        body: qs.stringify({
          grant_type: 'client_credentials',
        }),
        headers: {
          'content-type': formContentType,
        },
      },
    }) as AuthResponse;
    this.setBearerAuth(authResponse.access_token);
    return authResponse;
  }

  validateSignature(signature: string, body: any): boolean {
    if (!this.webhookSecret) {
      return true;
    }
    const bodyStr = JSON.stringify(body);
    const computedSignature = crypto
      .createHmac('sha256', this.webhookSecret as string)
      .update(bodyStr)
      .digest('hex');
    return signature === computedSignature;
  }

  async fetchVerification(url: string): Promise<VerificationResource> {
    return this.callHttp({ url }) as Promise<VerificationResource>;
  }

  async createIdentity(metadata?: Record<string, any>): Promise<IdentityResource> {
    return this.callHttp({
      path: 'v2/identities',
      requestOptions: {
        method: 'POST',
        body: metadata
          ? JSON.stringify({ metadata })
          : undefined,
      },
    }) as Promise<IdentityResource>;
  }

  private setClientAuth(clientId: string, clientSecret: string) {
    this.clientAuthHeader = `Basic ${isomorph.btoa(`${clientId}:${clientSecret}`)}`;
  }

  private setBearerAuth(accessToken: string) {
    this.bearerAuthHeader = `Bearer ${accessToken}`;
  }

  private async callHttp({
    path,
    url,
    requestOptions = {},
    authType = 'bearer',
  }: CallHttpParamsType) {
    let triedAuth = false;
    if (authType === 'bearer' && !this.bearerAuthHeader) {
      await this.auth();
      triedAuth = true;
    }
    if (authType !== 'none') {
      let authorization = null;
      if (authType === 'bearer') {
        authorization = this.bearerAuthHeader;
      } else if (authType === 'basic') {
        authorization = this.clientAuthHeader;
      }
      if (authorization) {
        const { headers = {} } = requestOptions;
        // @ts-ignore
        headers.authorization = authorization;
        requestOptions.headers = headers;
      }
    }

    const requestURL = url || `${this.host}/${path}`;
    try {
      return await callHttp(requestURL, requestOptions);
    } catch (err) {
      if (!triedAuth
        && authType === 'bearer'
        && err instanceof ErrorResponse
        && (err as ErrorResponse).response.status === 401
      ) {
        // re-auth
        await this.auth();
        // @ts-ignore
        requestOptions.headers.authorization = this.bearerAuthHeader;
        return callHttp(requestURL, requestOptions);
      }
      throw err;
    }
  }

  async sendInput(
    identityId: string,
    sendInputRequest: SendInputRequest,
  ): Promise<SendInputResponse> {
    // @ts-ignore
    const formData = new isomorph.FormData();
    formData.append('inputs', JSON.stringify(sendInputRequest.inputs));
    sendInputRequest.files.forEach((fileRecord) => {
      formData.append(fileRecord.mediaType, fileRecord.stream, fileRecord.fileName);
    });
    return this.callHttp({
      path: `v2/identities/${identityId}/send-input`,
      requestOptions: {
        method: 'POST',
        body: formData,
      },
    }) as Promise<SendInputResponse>;
  }
}

export default new ApiService();
