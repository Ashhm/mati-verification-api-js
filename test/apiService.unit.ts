import * as fs from 'fs';
import callHttp from '../src/lib/callHttp';
import { ApiService, Options } from '../src/apiService';
import WebhookResource, { EventNameTypes } from '../src/models/WebhookResource';
import AuthResponse from '../src/models/v2/AuthResponse';
import ErrorResponse from '../src/lib/ErrorResponse';
import IdentityResource, { IdentityStatusTypes } from '../src/models/v2/IdentityResource';
import SendInputResponse from '../src/models/v2/SendInputResponse';
import SendInputRequest, {
  DocumentPhotoInputData,
  DocumentTypeTypes,
  Input,
  InputTypeTypes,
  MediaTypeTypes,
  PageTypes,
} from '../src/models/v2/SendInputRequest';

jest.mock('../src/lib/callHttp');

const callHttpMock = callHttp as jest.Mock;

const clientId = 'clientId';
const clientSecret = 'clientSecret';
const webhookSecret = 'webhookSecret';

const authResponse: AuthResponse = {
  access_token: 'access_token',
  expiresIn: 3600,
  payload: { user: { _id: 'userId' } },
};

const METADATA = {
  email: 'john@gmail.com',
};

type MockFn = () => Promise<any>;
let authMock: MockFn | null;
let httpMock: MockFn | null;

callHttpMock.mockImplementation((requestURL: string): Promise<any> | null => {
  if (requestURL.endsWith('oauth')) {
    return authMock && authMock();
  } else {
    return httpMock && httpMock();
  }
});

type Context = { apiService: ApiService };

function createContext<C>(initialContext: Partial<C> = {}): C {
  return initialContext as C;
}
const withApiService = (context: Context, options: Options = {
  clientId,
  clientSecret,
  webhookSecret,
}) => {
  beforeAll(() => {
    context.apiService = new ApiService();
    context.apiService.init(options);
  });
  return context;
};

const withHttpMocks = (
  httpMockFn: MockFn = () => Promise.resolve(true),
  authMockFn: MockFn = () => Promise.resolve(authResponse),
) => {
  beforeAll(() => {
    authMock = authMockFn;
    httpMock = httpMockFn;
  });
  afterAll(() => {
    authMock = null;
    httpMock = null;
    callHttpMock.mockClear();
  });
};

describe('apiService', () => {
  describe('#init', () => {
    it('should not throw', () => {
      const apiService = new ApiService();
      expect(() => apiService.init({
        clientId,
        clientSecret,
      })).not.toThrow();
    });
  });

  describe('#validateSignature', () => {
    const webhookResource: WebhookResource = {
      eventName: EventNameTypes.VerificationCompleted,
      metadata: METADATA,
      resource: 'https://api.getmati.com/api/v1/verifications/db8d24783',
    };

    describe('when right signature', () => {
      const context: Context = withApiService(createContext());
      it('should return false', () => {
        expect(context.apiService.validateSignature(
          '0c5ed2cad914fd2a1571b47bb087953af574a353ff9d96f8603f8c0d7955340c',
          webhookResource,
        )).toBe(true);
      });
    });

    describe('when no webhookSecret', () => {
      const context: Context = withApiService(
        createContext(),
        {
          clientId,
          clientSecret,
        },
      );
      it('should return false', () => {
        expect(() => context.apiService.validateSignature('wrong sig', webhookResource)).toThrow();
      });
    });

    describe('when wrong signature', () => {
      const context: Context = withApiService(createContext());
      it('should return false', () => {
        expect(context.apiService.validateSignature('wrong sig', webhookResource)).toBe(false);
      });
    });
  });

  describe('#fetchResource', () => {
    const resourceUrl = 'http://resourceUrl';

    describe('with right auth', () => {
      type Resource = {
        name: string;
      };
      const resource: Resource = {
        name: 'resourceName',
      };
      const context: Context = withApiService(createContext());
      withHttpMocks(
        () => Promise.resolve(resource),
      );
      it('should resolve with resource',
        () => expect(context.apiService.fetchResource(resourceUrl)).resolves.toBe(resource));
      it('should call callHttp',
        () => expect(callHttpMock.mock.calls).toMatchSnapshot());
    });

    describe('with 401 error ', () => {
      const context: Context = withApiService(createContext());
      withHttpMocks(
        () => Promise.reject(new ErrorResponse('Oooups!', ({ status: 401 } as unknown) as Response)),
      );
      it('should resolve with resource',
        () => expect(context.apiService.fetchResource(resourceUrl))
          .rejects.toThrowErrorMatchingSnapshot());
      it('should call callHttp',
        () => expect(callHttpMock.mock.calls).toMatchSnapshot());
    });
  });

  describe('#createIdentity', () => {
    describe('without specified flowId', () => {
      const identityResource: IdentityResource = {
        _id: 'identityId',
        status: IdentityStatusTypes.Pending,
      };
      const context: Context = withApiService(createContext());
      withHttpMocks(
        () => Promise.resolve(identityResource),
      );
      it('should resolve with resource',
        () => expect(context.apiService.createIdentity(METADATA)).resolves.toBe(identityResource));
      it('should call callHttp',
        () => expect(callHttpMock.mock.calls).toMatchSnapshot());
    });

    describe('with specified flowId', () => {
      const flowId = '5ea832b35ada455758d3faa7';
      const identityResource: IdentityResource = {
        _id: 'identityId',
        status: IdentityStatusTypes.Pending,
      };
      const context: Context = withApiService(createContext(), {
        clientId,
        clientSecret,
        flowId,
      });
      withHttpMocks(
        () => Promise.resolve(identityResource),
      );
      it('should resolve with resource',
        () => expect(context.apiService.createIdentity(METADATA)).resolves.toBe(identityResource));
      it('should call callHttp',
        () => expect(callHttpMock.mock.calls).toMatchSnapshot());
    });
  });

  describe('#sendInput', () => {
    enum FileNames {
      Front = 'front.png',
      Video = 'video.mp4',
    }

    const sendInputRequest: SendInputRequest = {
      inputs: [
        <Input<DocumentPhotoInputData>>{
          inputType: InputTypeTypes.DocumentPhoto,
          group: 0,
          data: {
            type: DocumentTypeTypes.NationalId,
            country: 'US',
            region: 'IL',
            page: PageTypes.Front,
            filename: FileNames.Front,
          },
        },
      ],
      files: [
        {
          mediaType: MediaTypeTypes.Document,
          fileName: FileNames.Front,
          stream: fs.createReadStream(`./assets/${FileNames.Front}`),
        },
        {
          mediaType: MediaTypeTypes.Video,
          fileName: FileNames.Video,
          stream: fs.createReadStream(`./assets/${FileNames.Video}`),
        },
      ],
    };
    const sendInputResponse: SendInputResponse = [
      { result: true },
      { result: false },
    ];
    const context: Context = withApiService(createContext());
    withHttpMocks(() => Promise.resolve(sendInputResponse));
    it('should resolve with resource',
      () => expect(context.apiService.sendInput('identityId', sendInputRequest))
        .resolves.toBe(sendInputResponse));
    it('should call callHttp', () => {
      callHttpMock.mock.calls[1][1].body = {};
      expect(callHttpMock.mock.calls).toMatchSnapshot();
    });
  });
});
