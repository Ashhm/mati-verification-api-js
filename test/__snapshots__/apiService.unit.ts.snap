// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiService #createIdentity should call callHttp 1`] = `
Array [
  Array [
    "https://api.getmati.com/oauth",
    Object {
      "body": "grant_type=client_credentials",
      "headers": Object {
        "authorization": "Basic Y2xpZW50SWQ6Y2xpZW50U2VjcmV0",
        "content-type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://api.getmati.com/v2/identities",
    Object {
      "body": Object {
        "metadata": Object {
          "email": "john@gmail.com",
        },
      },
      "headers": Object {
        "authorization": "Bearer access_token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`apiService #fetchResource with 401 error  should call callHttp 1`] = `
Array [
  Array [
    "https://api.getmati.com/oauth",
    Object {
      "body": "grant_type=client_credentials",
      "headers": Object {
        "authorization": "Basic Y2xpZW50SWQ6Y2xpZW50U2VjcmV0",
        "content-type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
  Array [
    "http://resourceUrl",
    Object {
      "headers": Object {
        "authorization": "Bearer access_token",
      },
    },
  ],
]
`;

exports[`apiService #fetchResource with 401 error  should resolve with resource 1`] = `"Oooups!"`;

exports[`apiService #fetchResource with right auth should call callHttp 1`] = `
Array [
  Array [
    "https://api.getmati.com/oauth",
    Object {
      "body": "grant_type=client_credentials",
      "headers": Object {
        "authorization": "Basic Y2xpZW50SWQ6Y2xpZW50U2VjcmV0",
        "content-type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
  Array [
    "http://resourceUrl",
    Object {
      "headers": Object {
        "authorization": "Bearer access_token",
      },
    },
  ],
]
`;

exports[`apiService #sendInput should call callHttp 1`] = `
Array [
  Array [
    "https://api.getmati.com/oauth",
    Object {
      "body": "grant_type=client_credentials",
      "headers": Object {
        "authorization": "Basic Y2xpZW50SWQ6Y2xpZW50U2VjcmV0",
        "content-type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://api.getmati.com/v2/identities/identityId/send-input",
    Object {
      "body": Object {},
      "headers": Object {
        "authorization": "Bearer access_token",
      },
      "method": "POST",
    },
  ],
]
`;
